### DISCUSSION TITLE
Abstracted Architecture

### DISCUSSION THREAD

**Discussion by meganindya** [sugarlabs/musicblocks-v4#2] [Ideas]
Identified Requirements:

| Engine | Client |
| -------- | ------- |
| <ul><li>[ ] State Management</li></ul> | <ul><li>[ ] Canvas Context</li></ul> |
| <ul><li>[ ] Scheduling</li></ul> | <ul><li>[ ] Audio Context</li></ul> |
| <ul><li>[ ] Block Framework</li></ul> | <ul><li>[ ] Block Tree</li></ul> |
| <ul><li>[ ] Feature API</li></ul> | <ul><li>[ ] Block Artwork</li></ul> |
| <ul><li>[ ] Export Algorithms</li></ul> | <ul><li>[ ] Block Interaction</li></ul> |
|  | <ul><li>[ ] ADSR Envelope</li></ul> |
|  | <ul><li>[ ] History</li></ul> |
|  | <ul><li>[ ] Context Save Logic</li></ul> |
|  | <ul><li>[ ] Widgets</li></ul> |

**Comment by meganindya** [comment]
This is a very abstract representation of the requirements.

https://lucid.app/documents/view/cbb01bc9-9407-4aa9-8173-fccb64b1ee68

![Abstracted Architecture](https://user-images.githubusercontent.com/33328728/101639061-dc24df80-3a54-11eb-9beb-43ac68ecd083.png)

**Comment by walterbender** [comment]
What is going on with the ADSR envelop?

**Comment by meganindya** [comment]
Tone.js provides functionality to trigger an attack and a release separately, or at once with `triggerAttackRelease`, which works on precise musical time. There's also functionality for `PolySynth` which allows multiple voices to be managed together.

IG if we can process the required sound envelope properties in the "engine", perhaps a `AudioContext` wrapper in the "client" can effectively dispatch the events via the Tone.js API — separation of concerns.

**Comment by walterbender** [comment]
We use all of that functionality in musicsynths.js. But I think I see where you are going. It would also give us more direct control over sustaining notes and "bending" notes.

**Comment by meganindya** [comment]
My original idea was that the "engine" should be independent of Tone.js. But, I think it's best if we use `Tone.Transport`  for timing and scheduling.

**Comment by walterbender** [comment]
I completely agree. Should help a great deal with performance and robustness.

**Comment by walterbender** [comment]
And it will make our code less complex.

**Comment by meganindya** [comment]
This has evolved to:

![Music Blocks v4 - component architecture](https://user-images.githubusercontent.com/33328728/140659039-af3ae263-d143-4b5d-84b4-1eed4ba8a4ec.jpg)
